File: algorithm/Programmers/Level 1/python/두 개 뽑아서 더하기.py
# itertools 라이브러리를 이용한 풀이와 dfs 백트래킹을 이용해 직접 조합을 구한 풀이다.

from itertools import combinations
def solution(numbers):
    answer = set()
    for i in combinations(numbers,2):
        answer.add(sum(i))
    answer = sorted(list(answer))
    return answer

# def dfs(cnt,idx,numbers,ls,answer):
#     if cnt==2:
#         answer.add(sum(ls))
#         return
#     for i in range(idx,len(numbers)):
#         ls.append(numbers[i])
#         dfs(cnt+1,i+1,numbers,ls,answer)
#         ls.pop()

# def solution(numbers):
#     answer = set()
#     ls = []
#     dfs(0,0,numbers,ls,answer)
#     answer = sorted(answer)
#     return answer

# print(solution([2,1,3,4,1]))

File: algorithm/Programmers/Level 1/javascript/두 개 뽑아서 더하기.js
// array.indexOf(element)는 array에서 element의 인덱스를 반환한다. 해당 element가 없으면 -1을 반환한다.
// array.sort()는 유니코드 기준 정렬(사전식 정렬), 수의 오름차순 정렬을 하기 위해서는 sort((a,b)=>a-b)를 해야한다.

function solution(numbers) {
    var answer = []
    for (var i = 0; i < numbers.length - 1; i++) {
        for (var j = i + 1; j < numbers.length; j++) {
            if (answer.indexOf(numbers[i] + numbers[j]) == -1) {
                answer.push(numbers[i] + numbers[j]);
            }
        }
    }
    answer.sort((a, b) => a - b);
    return answer;
}
console.log(solution([5, 0, 2, 7]));